internal/wal_push_handler.go

```diff
@@ -58,17 +58,19 @@ func (err CantOverwriteWalFileError) Error() string {
 // TODO : unit tests
 // HandleWALPush is invoked to perform wal-g wal-push
 func HandleWALPush(uploader *WalUploader, walFilePath string) {
+	uploader.UploadingFolder = uploader.UploadingFolder.GetSubFolder(utility.WalPath)
 	if uploader.ArchiveStatusManager.IsWalAlreadyUploaded(walFilePath) {
 		err := uploader.ArchiveStatusManager.UnmarkWalFile(walFilePath)
 
 		if err != nil {
 			tracelog.ErrorLogger.Printf("unmark wal-g status for %s file failed due following error %+v", walFilePath, err)
 		}
+		if viper.GetString(UploadWalMetadata) == WalBulkMetadataLevel && walFilePath[len(walFilePath)-1:] == "F" {
```

Comment:
This looks identical to the line 85.

I think that it might be better to move the whole metadata upload logic to the `WalUploader` and upload the metadata files similarly to WAL delta files.

Reply:
Also, what if the WAL segment ending with "F" will be uploaded faster than the previous segments?

Reply:
> This looks identical to the line 85.
> 
> Maybe extract the common logic to avoid duplication? Or even move the whole metadata upload logic to the `WalUploader` and upload the metadata files similarly to WAL delta files.

Thanks. Just for my understanding, to move the metadata logic to https://github.com/wal-g/wal-g/blob/master/internal/wal_uploader.go#L45 ?
wal_receiver_handler.go also uses the UploadWalFile@wal_uploader.go, should it have any clash if the metadata logic is moved there.

Reply:
> Also, what if the WAL segment ending with "F" will be uploaded faster than the previous segments? Will the current bulk uploading logic work?

The logic was added here https://github.com/wal-g/wal-g/pull/899/files#diff-f66d6177f62076d018bb01e9b1e9056c991e8206cdb9d16db9e817faf4356abfR68 for that purpose only.
The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks

Reply:
> wal_receiver_handler.go also uses the [UploadWalFile@wal_uploader.go](mailto:UploadWalFile@wal_uploader.go), should it have any clash if the metadata logic is moved there.

Nice catch, I think that it might be a good idea to upload the WAL metadata during wal-receive too.

> The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks

When `WALG_UPLOAD_CONCURRENCY` is higher than 1, WAL-G invokes the BgUploader, which looks for WAL files adjacent to the current WAL file and uploads them concurrently to the storage. So, on a single archive_command call there could be multiple WALs being uploaded.

Reply:
> > wal_receiver_handler.go also uses the [UploadWalFile@wal_uploader.go](mailto:UploadWalFile@wal_uploader.go), should it have any clash if the metadata logic is moved there.
> 
> Nice catch, I think that it might be a good idea to upload the WAL metadata during wal-receive too. So I guess that becomes one more reason to move the metadata uploading to the wal_uploader.go
> 
> > The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks
> 
> When `WALG_UPLOAD_CONCURRENCY` is higher than 1, WAL-G invokes the BgUploader, which looks for WAL files adjacent to the current WAL file and uploads them concurrently to the storage. So, on a single archive_command call there could be multiple WALs being uploaded. So there is no guarantee that the "F" file will be uploaded exactly after the previous WAL files.

Thanks. I will move the logic to wal_uploader.go so that wal-push and wal-receive can use the metadata logic.